core/services/relay/evm/chain_reader.go

```diff
@@ -197,6 +198,26 @@ func (cr *chainReader) QueryKey(ctx context.Context, contractName string, filter
 	return b.QueryKey(ctx, filter, limitAndSort, sequenceDataType)
 }
 
+func (cr *chainReader) ReplaySequence(ctx context.Context, contractName, key, blockID string) error {
+	binding, err := cr.bindings.GetReadBinding(contractName, key)
+	if err != nil {
+		return err
+	}
+
+	_, isEventBinding := binding.(*eventBinding)
+	if !isEventBinding {
+		return fmt.Errorf("key: %s does not belong to a replayable event", key)
+	}
+
+	blockNumber, err := strconv.ParseInt(blockID, 10, 64)
+	if err != nil {
+		return fmt.Errorf("%w: EVM blockID should be a number but is %s", err, blockID)
```

Comment:
Why wrap the source error as a prefix instead of the standard suffix? Seems confusing to read.
```suggestion
		return fmt.Errorf("failed to parse EVM blockID as number %q: %w", blockID, err)
```

Reply:
I think we should not expose this and allow calling it directly. Is there some way we can call it from `(e *eventBinding) Register`?

I'm a little unclear on the relationship between `Bind` and `Register`, is `Register()` only called from `Bind` when it's called, or can you call `Register()` and `Unregister()` dynamically after you've bound a contract?

We've talked before about some products needing ability to register and unregister sequences dynamically, but I can't recall where we landed--is that considered "planned for the future" rather than a part of the current implementation?

Also, I just noticed this while looking over the code, is this a bug? (only calls `e.Register()` if `e.registerCalled` is already true?)

```
  func (e *eventBinding) Bind(ctx context.Context, binding commontypes.BoundContract) error {
      // it's enough to just lock bound here since Register/Unregister are only called from here and from Start/Close
      // even if they somehow happen at the same time it will be fine because of filter lock and hasFilter check
      e.bindLock.Lock()
      defer e.bindLock.Unlock()

      if e.bound {
          // we are changing contract address reference, so we need to unregister old filter it exists
          if err := e.Unregister(ctx); err != nil {
              return err
          }
      }

      e.address = common.HexToAddress(binding.Address)

      // filterRegisterer isn't required here because the event can also be polled for by the contractBinding common filter.
      if e.filterRegisterer != nil {
          id := fmt.Sprintf("%s.%s.%s", e.contractName, e.eventName, uuid.NewString())
          e.pollingFilter.Name = logpoller.FilterName(id, e.address)
          e.pollingFilter.Addresses = evmtypes.AddressArray{e.address}
          e.bound = true
          if e.registerCalled {
              return e.Register(ctx)
          }
      }
      e.bound = true
      return nil
  }```
  
Anyway, I think we ought to add `startBlockID` as a parameter to `Register`, and possibly also `Bind` if that's the only time it's called. Basically any time `lp.Register()` is called, `lp.Async()` should be called immediately after that. (Unless `startBlockID` is nil I guess.)

More details on why exposing this directly would be bad:

If we allow callers to do anything outside of that which applies to some events but not others, then when `QueryKey()` is called for a particular block range, it would have to give back a mix of some events missing and others not missing, depending on which ones happen to have been downloaded by that point. None of the existing rules we have for consistency would be enforced, for example:  we could no longer guarantee that there are no gaps where we have logs for some blocks, then are missing logs for a while, then have more logs for a higher block range.  There could end up being any number of disconnected log events floating around with no way to tell what's going on from the outside. Also, the implementation would need to make LogPoller a lot more complex, adding a column to its filter table to track the progress of each filter individually, as to which logs have been downloaded. Currently we have a `LatestBlock` method that tells callers what the most recent block is that LogPoller guarantees it has logs for--this would no longer be useful since it would be a different block for every event type.

Also, we should only ever call `lp.ReplayAsync`, as it's important that the replay is complete before a particular filter is registered. Until it's complete, we can't return from `Bind()`, because then the caller won't know whether it's safe to call `QueryKey` since whether the logs are ready yet for a particular block would just be up to an arbitrary race condition, ie `LatestBlock` would return inaccurate results for that event.

Reply:
Context: there is a plan to add `fromBlock` param to `lp.RegisterFilter()`, so that it doesn't have to do a global replay. Then we can disallow any plugins from calling Replay directly, and only allow that from the CLI.

Reply:
> I'm a little unclear on the relationship between Bind and Register, is Register() only called from Bind when it's called, or can you call Register() and Unregister() dynamically after you've bound a contract?

Reply:
> I'm a little unclear on the relationship between Bind and Register, is Register() only called from Bind when it's called, or can you call Register() and Unregister() dynamically after you've bound a contract?
> 
> We've talked before about some products needing ability to register and unregister sequences dynamically, but I can't recall where we landed--is that considered "planned for the future" rather than a part of the current implementation?
> 
> Also, I just noticed this while looking over the code, is this a bug? (only calls e.Register() if e.registerCalled is already true?)

There are comments in the code, it would be helpful if you could step through the code a bit before commenting to make these discussions more efficient. 
Bind calls register only if it was called at least once before because:
```
// registerCalled is used to determine if Register was called during Chain Reader service Start.
// This is done to avoid calling Register while the service is not running because log poller is most likely also not running.
	registerCalled bool
```
This also allows us to Bind contracts later to a new address which will register new filters

Reply:
> Anyway, I think we ought to add startBlockID as a parameter to Register, and possibly also Bind if that's the only time it's called. Basically any time lp.Register() is called, lp.Async() should be called immediately after that. (Unless startBlockID is nil I guess.)
This is not possible since we aren't exposing  `Log Poller` filter registering (filters are kept in the config), and adding this parameter to `Bind` makes no sense since `Bind` binds the whole contract so you would be setting starting block for every event in that contract.